lib/Elevate/Components/CCS.pm

```diff
@@ -0,0 +1,461 @@
+                WHERE addressbook_home.owner_uid = ?
+                AND addressbook_object.is_in_trash = ?;",
+        },
+    );
+
+    return ( $querydata{$type}{'query'}, $querydata{$type}{'args'} );
+}
+
+sub _get_dbh ($self) {
+    $self->{dbh} ||= Cpanel::CCS::DBUtils::get_dbh();
+    return $self->{dbh};
+}
+
+sub _get_export_path_for_user ( $self, $user ) {
+    $self->{$user}{export_path} ||= EXPORT_DIR() . '/' . $user . '/calendar_and_contacts';
+    return $self->{$user}{export_path};
+}
+
+sub _get_ccs_info_for_user ( $self, $user ) {
+    $self->{$user}{info} ||= Cpanel::CCS::Userdata::get_cpanel_account_users_uuids($user);
+    return $self->{$user}{info};
+}
+
+sub _ensure_export_directory ($self) {
+    File::Path::make_path(EXPORT_DIR);
+
+    # Do not use File::Path to do this since it only
+    # runs chmod if it creates the directory and we want
+    # to make sure it is 0700 regardless if it was just created
+    # or not
+    chmod 0700, EXPORT_DIR;
+
+    return;
+}
+
+####################################
+##### post_leapp code below this ###
+####################################
+
+sub post_leapp ($self) {
+    return unless Elevate::StageFile::read_stage_file('ccs_installed');
+
+    $self->_install_ccs_and_dependencies();
+
+    # This needs to happen before verifying that the service is up
+    # There is a task created that makes a schema update that can
+    # cause CCS to fail to start
+    $self->_clear_task_queue();
+
+    $self->_ensure_ccs_service_is_up();
+    $self->run_once('import_ccs_data');
+
+    return;
+}
+
+sub _install_ccs_and_dependencies ($self) {
+    my @packages_to_install = ( CCS_PACKAGE() );
+
+    push @packages_to_install, ZPUSH_PACKAGE() if Elevate::StageFile::read_stage_file('zpush_installed');
+
+    $self->dnf->install(@packages_to_install);
+
+    return;
+}
+
+sub _clear_task_queue ($self) {
+
+    # CCS queues up tasks when it is first installed
+    # These need to run before we attempt to import data
+    $self->ssystem( TASK_QUEUE_SCRIPT, 'run' );
+    return;
+}
+
+sub _ensure_ccs_service_is_up ($self) {
+
+    INFO('Attempting to ensure that the CCS service is running');
+
+    my $attempts     = 1;
+    my $max_attempts = 5;
+    while ( $attempts <= $max_attempts ) {
+        DEBUG("Attempt $attempts of $max_attempts to verify that the CCS service is up");
+
+        if ( $self->_ccs_service_is_up() ) {
+            INFO('Verified that the CCS service is up');
+            return;
+        }
+
+        # If the service was not up at this point, it is likely that
+        # the schema update failed during the install.  In this case,
+        # the only way I have found to get things working again is to
+        # completely remove the package and start the install over
+
+        $self->remove_cpanel_ccs_home_directory();
+        $self->remove_cpanel_ccs_home_directory();
```

Comment:
Redundant?

Reply:
Sigh, copy/paste error.  Line 377 should be removing the package